### DISCUSSION TITLE
Unsupervised Relation Extraction

### DISCUSSION THREAD

**Discussion by gonzobrandon** [explosion/spaCy#7449] [Help: Best practices]
I see the great tutorial from @svlandeg  on relation extraction - where (to my understanding) we supply a categorical list of relations to train and then predict against. For example: `{binding, inhibition, activation}` as a list of possible relations in the video example. 

How does this method scale to something like thousands of categories or with one-shot learning? What I'm aiming to try is to find a relation between two (NER) entities with either no pre-knowledge of the linguistic relationship or a subset of corpus tokens nearby to query against. An example: _"Today, Kevin Quinn won the New York City marathon"_ ...in this case the word "won" would be the specific relationship I'm after. But what if the training model has not seen this specific relationship (verb, in this case). I believe this is called Open-IE-style. I could reduce the relations down to perhaps 1,000-5,000 key phrases and words if needed. Certainly, calculating token distance via embeddings can be done for out-of-vocab or out-of-corpus words.

Any papers or implementations would be very appreciated. If any engineers or consultants are available to work on this, please reach out.

**Comment by svlandeg** [comment]
Hi!

I got confused with the terminology of your post for a second, so just to clarify, within spaCy code & docs, we define:
- `entity linking` as the process of linking a textual mention (e.g. "President Obama") to a unique database identifier, e.g. [`"Q76"`](https://www.wikidata.org/wiki/Q76).
- `relation extraction` as the process of determining whether or not two (or more) entities are in a semantic relation as expressed by the sentence context, and if so, labeling that relation.

I understand your question is about `relation extraction`, though the title threw me off ;-)

Anyway, all that to then give you a disappointing answer that personally, I haven't really looked into this "open ended" style of relation extraction yet. But I'd be interested in hearing from others in the community if they have any relevant pointers!

**Reply by gonzobrandon** [reply]
Thanks @svlandeg. Yes, the title was off (fixing).  Ultimately my pipeline includes NER -> NEL  and the last step is indeed the "open ended" relation extraction. 

I have had some success with implementing  the Hearst Patterns. Some of the sources I turned to were:

[1]  [Hearst Patterns Revisited: Automatic Hypernym Detection from Large Text Corpora](https://research.fb.com/wp-content/uploads/2018/07/Hearst-Patterns-Revisited-Automatic-Hypernym-Detection-from-Large-Text-Corpora.pdf) seems to have had a similar experience to my own. 

[2] [Hearst Patterns with Spacy pip package (mmichelsonIF/hearst_patterns_python)](https://github.com/mmichelsonIF/hearst_patterns_python) 

[3] [Introduction to Information Extraction using Python and spaCy](https://www.analyticsvidhya.com/blog/2019/09/introduction-information-extraction-python-spacy/) A good overview of some approaches. (Thanks Walid from UBIAI)

Some implementation thoughts on improving on the rule/pattern approach:

1. *Training a model*:   I would naturally like to train a neural network with spacy/prodigy that could take some number of example documents and like was done in the Emerson example, as input provide `{Sentence Context, Sentence Dependency Tree, POS Tags}` from which perhaps there is enough of a pattern to generalize. This is what I am working on. I have yet to implement the annotation tool in either Prodigy or another tool, such as 
[UBIAI](https://ubiai.tools) but I will start this very soon.

2.  *Extending Emerson Project*: I thought about a very large extension of the Emerson project and including thousands or tens of thousands of token/phrases that summarize the relation well, but annotating and scaling that model might be challenging to say the least.


3. I think the open-ended RE problem is tractable because a given sentence context (or some *n* number of sentences before) contain the verb or adjective token or span that contain the best relation...So, we're not searching from zero-knowledge or trying every possible relation possible in the language. 

4. There is a lot of research and datasets available for extracting triples. Some of them try to restrict the relation to a hybrid identifier that can be adapted to OWL/RDF data format such as `has_buyer` etc - however could be equally useful to extract a natural language representation.

    * [Borcan. Python NLP Tutorial: Information Extraction and Knowledge Graphs (2020)](https://medium.com/swlh/python-nlp-tutorial-information-extraction-and-knowledge-graphs-43a2a4c4556c)

    * [Sakor et. al. Old is Gold: Linguistic Driven Approach for Entity and Relation Linking of Short Text (2019)](https://www.aclweb.org/anthology/N19-1243/) 

    * [Gozwami et.al. Unsupervised Relation Extraction from Language Models using Constrained Cloze Completion (2020)](https://arxiv.org/abs/2010.06804)  

Wold love community input on this topic. Thank you!
